cclib/method/bickelhaupt.py

```diff
@@ -0,0 +1,122 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (c) 2020, the cclib development team
+#
+# This file is part of cclib (http://cclib.github.io) and is distributed under
+# the terms of the BSD 3-Clause License.
+
+"""Calculation of Bickelhaupt population analysis based on data parsed by cclib."""
+
+import random
+
+import numpy
+
+from cclib.method.population import Population
+
+
+class Bickelhaupt(Population):
+    """Bickelhaupt population analysis."""
+
+    def __init__(self, *args):
+
+        # Call the __init__ method of the superclass.
+        super(Bickelhaupt, self).__init__(logname="Bickelhaupt Population Analysis", *args)
+
+    def __str__(self):
+        """Return a string representation of the object."""
+        return "Bickelhaupt charges of {}".format(self.data)
+
+    def __repr__(self):
+        """Return a representation of the object."""
+        return "Bickelhaupt({})".format(self.data)
+
+    def calculate(self, indices=None, fupdate=0.05):
+        """Perform a Bickelhaupt population analysis."""
+
+        # Determine number of steps, and whether process involves beta orbitals.
+        self.logger.info("Creating attribute aoresults: [array[2]]")
+        nbasis = self.data.nbasis
+        alpha = len(self.data.mocoeffs[0])
+        self.aoresults = [ numpy.zeros([alpha, nbasis], "d") ]
+        w = [ numpy.zeros([self.data.nbasis, self.data.nbasis], "d") ]
+        nstep = alpha + nbasis
+        unrestricted = (len(self.data.mocoeffs) == 2)
+        if unrestricted:
+            beta = len(self.data.mocoeffs[1])
+            self.aoresults.append(numpy.zeros([beta, nbasis], "d"))
+            w.append(numpy.zeros([self.data.nbasis, self.data.nbasis], "d"))
+            nstep += (beta + nbasis)
+
+        # Intialize progress if available.
+        if self.progress:
+            self.progress.initialize(nstep)
+
+        step = 0
+
+        # First determine the weight matrix
+        for spin in range(len(self.data.mocoeffs)):
+            for i in range(self.data.nbasis):
+                if self.progress and random.random() < fupdate:
+                    self.progress.update(step, "Bickelhaupt Population Analysis")
+                for j in range(self.data.nbasis):
+                    # W_{ab} = 2 * \sum_k c_{ak}^2 / (\sum_i c_{ai}^2 + \sum_j c_{bj}^2)

```

Comment:
Perhaps move this to the comment at the top of the block, and perhaps also add a short description of the weight matrix.